src/DotLiquid/Context.cs

```diff
@@ -522,67 +522,50 @@ private bool TryGetVariable(IEnumerator<string> partsEnumerator, out object vari
-                        object res = ((KeyValuePair<string, object>)@object).Value;
-                        @object = Liquidize(res);
+                        @object = castCollection.Count();
                     }
-                    // If object is a hash- or array-like object we look for the
-                    // presence of the key and if its available we return it
-                    else if (TryEvaluateHashOrArrayLikeObject(@object, part, out var hashObj))
+                    else if (Template.NamingConvention.OperatorEquals(part as string, "first"))
                     {
-                        // If its a proc we will replace the entry with the proc
-                        @object = Liquidize(hashObj);
-                    }
-                    // Some special cases. If the part wasn't in square brackets and
-                    // no key with the same name was found we interpret following calls
-                    // as commands and call them on the current object
-                    else if (!partResolved && (@object is IEnumerable) && (Template.NamingConvention.OperatorEquals(part as string, "size") || Template.NamingConvention.OperatorEquals(part as string, "first") || Template.NamingConvention.OperatorEquals(part as string, "last")))
-                    {
-                        var castCollection = ((IEnumerable)@object).Cast<object>();
-                        if (Template.NamingConvention.OperatorEquals(part as string, "size"))
-                            @object = castCollection.Count();
-                        else if (Template.NamingConvention.OperatorEquals(part as string, "first"))
-                        {
-                            object res = castCollection.FirstOrDefault();
-                            @object = Liquidize(res);
-                        }
-                        else if (Template.NamingConvention.OperatorEquals(part as string, "last"))
-                        {
-                            object res = castCollection.LastOrDefault();
-                            @object = Liquidize(res);
-                        }
+                        @object = Liquidize(castCollection.FirstOrDefault());
                     }
-                    // No key was present with the desired value and it wasn't one of the directly supported
-                    // keywords either. The only thing we got left is to return nil
                     else
                     {
-                        variable = null;
-                        return false;
+                        @object = Liquidize(castCollection.LastOrDefault());
                     }
                 }
+                // No key was present with the desired value and it wasn't one of the directly supported
+                // keywords either. The only thing we got left is to return nil
+                else
+                {
+                    variable = null;
+                    return false;
+                }

```

Comment:
Not sure what this code change has to do with the this PR, please revert or explain

Reply:
These changes were made for pre 2.2 combability before implementing the LegacyKeyValueDrop.
I retained them because they:
- update misleading comments to reflect the current code behavior (several comments appeared to be cut/pasted in error).
- merged if statements originally on l535 and l547 into a single block to remove extra `is KeyValuePair`.
- remove a redundant else block (l543).
- remove unreachable branch (l572) caused by the if statement (l562 ensure values can only be size/first/last).
- remove unused variable assigns (l530, 538, 569, 574)

Reply:
I reviewed in a compare tool. Much easier to understand. Marking resolved

Reply:
I reviewed in a compare tool. Much easier to understand. Marking resolved

Reply:
I reviewed in a compare tool. Much easier to understand. Marking resolved